File: rust/src/arithmetic_variable_node.rs
use std::collections::HashMap;
use crate::arithmetic_type::ArithmeticType;
use crate::calculation_error::CalculationError;
use crate::generic_node::GenericNode;
use crate::instruction_serialiser::ArithmeticResultNodeWrapper;
use crate::instruction_serialiser::arithmetic_result_node_wrapper::Node;
use crate::parameter::Parameter;

pub use crate::instruction_serialiser::ArithmeticVariableNode;

impl GenericNode<ArithmeticType> for ArithmeticVariableNode {
    fn calculate(&self, parameters: &HashMap<&str, Parameter>) -> Result<ArithmeticType, CalculationError> {
        match parameters.get(self.variable_name.as_str()).ok_or(
            CalculationError::new(
                format!("The instruction expected an arithmetic parameter '{}' to be passed", self.variable_name).as_str()
            )
        )? {
            Parameter::Arithmetic(param) => Ok(*param),
            _ => Err(
                CalculationError::new(
                    format!("The instruction expected parameter '{}' to be arithmetic", self.variable_name).as_str()
                )
            )
        }
    }
}

impl ArithmeticVariableNode {
    pub fn new(variable_name: &str) -> ArithmeticResultNodeWrapper {
        ArithmeticResultNodeWrapper{
            node: Some(
                Node::VariableNode(
                    ArithmeticVariableNode{
                        variable_name: String::from(variable_name)
                    }
                )
            )
        }
    }
}

File: typescript/src/arithmetic_variable_node.ts
import { ArithmeticVariableNode } from './instruction_serialiser/instruction_serialiser_pb';
import { CalculationError } from './calculation_error';
import { CalculatableNode } from './calculatable_node';
import { ArithmeticResultNodeWrapper } from './arithmetic_result_node_wrapper';
import { InstructionParameters } from './instruction_parameters';

declare module './instruction_serialiser/instruction_serialiser_pb' {
  interface ArithmeticVariableNode extends CalculatableNode<number> {}
}

ArithmeticVariableNode.prototype.calculate = function(parameters: InstructionParameters): number {
  const val = parameters[this.getVariableName()];
  if (val === undefined) {
    throw new CalculationError(`The variable ${this.getVariableName()} was not passed`);
  } else if (Number.isFinite(val)) {
    return val as number;
  }
  throw new CalculationError(`The variable ${this.getVariableName()} needs to be of type number`);
};

function createArithmeticVariableNode(variableName: string): ArithmeticResultNodeWrapper {
  return new ArithmeticResultNodeWrapper().setVariableNode(
    new ArithmeticVariableNode().setVariableName(variableName)
  );
}

export { ArithmeticVariableNode, createArithmeticVariableNode };
